### PR TITLE
Ruby 2: my_enumerables

### PR DESCRIPTION
**Pull request by mke2111**
In this project, I recreated the ruby enumerable methods to serve the same purpose as the original methods.

They include:

1. my_each
2. my_each_with_index
3. my_all?
4. my_none?
5. my_any?
6. my_map
7. my_select
8. my_inject
9. my_count

### CODE REVIEW

File: README.md
Comment (teekaytech) [inline README.md]: Since there are no test cases for this project, you can remove this section of your README.md file.

File: README.md
Comment (teekaytech) [inline README.md]: It looks like you're the only contributor to this project. Kindly change `Check our code` to `Check my code`

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__This is an excellent job. However, you have to make some changes before this project can be approved. I will try to break things down as much as I can so that you can understand what each error means, and find a way to solve them without affecting the current flow.  Let's dive in.__

**[ REQUIRED CHANGES ]**

- [x] **Your GitHub Actions linters are not working.**
- This is because you nested the linters configuration file (`.github/workflows/linters.yml`) inside the `lib` directory. You should add the configuration file in the root directory (`Enumerables folder`) of your project.
- Ensure that all tests are passed (i.e. you have green marks) on your PR page before submitting another review request.

- [x] **`my_each` method failed some test cases.**
- If no block is given, `my_each` should return an enumerator. For instance,

```
 [0,1,2,3,4].each  # returns #<Enumerator: [0, 1, 2, 3, 4]:each>
[0,1,2,3,4].my_each  # returns no block given (yield) (LocalJumpError)
```
- When a block is given and the `my_each` method is called on a **range of values**, it should return the range itself after calling the given block once for each element in the range. For instance,

```
(1..4).each{ |num| num < 8 / 2 }   # returns 1..4 (the range itself)
(1..4).my_each{ |num| num < 8 / 2 }  # returns [1, 2, 3, 4]
```

**You are having these two similar error cases for your `my_each_with_index` method, and the first error case for your `my_select` method also. Please use the above scenarios as applicable.**

- [x] **`my_all?` method failed a test case.**
- When no block or argument is given, `my_all?` should return true when none of the collection members are false or nil, and false otherwise. For instance, 

```
[1, true, 'hi', []].all?  # returns true
[1, true, 'hi', []].my_all?  # returns false
```

- [x] **`my_any?` method failed a test case.**
- When no block or argument is given, `my_any?` should return false if at least one of the collection is not true. For Instance, 

```
[nil, false, nil, false].any?  # returns false
[nil, false, nil, false].my_any?  # returns true
```

- [x] **`my_none?` method failed a test case.**
- When no block or argument is given, `my_none?` should return true only if none of the collection members is true. For Instance, 

```
[nil, false, nil, false].none?  # returns true
[nil, false, nil, false].my_none?  # returns false
```

- [x] **`my_count` method failed a test case.**
- When called on a range, `my_count` should return the number of items in the range through enumeration::range. For Instance, 

```
(1..10).count  # returns 10
(1..10).my_count # returns "undefined method `length' for 1..10:Range (NoMethodError)"
```
From the error message, it is clear that the `.length` method cannot be called on a range. Find a way to create this exception so that your method returns the number of values in a range, without affecting the current flow.

- [x] **`my_inject` method failed a test case.**
- When no block or argument is given, `my_inject` should raise a `LocalJumpError` but no error is raised. For instance, 

[nil, false, nil, false].my_inject  # returns nil (No error raised)
[nil, false, nil, false].inject  # returns 'no block is given (LocalJumpError)' (error raised).

**[ OPTIONAL CHANGES ]**

- [x] It's good that you test your module with a test script (`run_test.rb`). However, I personally think it's not necessary you commit this file. You can remove it, add it to your `.gitignore` file, and re-create it so that it does not get committed.

**[ COMMENTS ]**
- You've come a long way in this project. Do not be discouraged by these changes.
- Find ways to make these changes without tampering with the current flow. This is very important because if care is not taken, a change may cause failure for some of the already passed test cases. You can use conditionals (if statement) to create different ways of handling different scenarios and data structures. All the best!

* * * 
**Please make the required changes and submit another review request when you are done.**

* * *

**_You can reach out to me on Twitter (@ola_lere), Github(@teekaytech), or slack (@taofeekolalere )_**
__Happy Coding :smiley:__

File: lib/main.rb
Comment (maelfosso) [inline lib/main.rb]:
```rb
2.6.5 :014 > block = proc { |num| num < 4 }
2.6.5 :015 > array
 => [2, 4, 3, 7, 1, 3, 1, 0, 6, 0, 4, 4, 3, 0, 3, 8, 6, 7, 8, 0, 5, 3, 2, 3, 2, 6, 5, 6, 2, 7, 6, 4, 3, 1, 0, 4, 0, 6, 6, 3, 0, 2, 0, 6, 8, 2, 2, 8, 3, 6, 4, 2, 1, 1, 4, 4, 3, 2, 8, 1, 6, 5, 4, 0, 6, 7, 5, 4, 8, 2, 2, 8, 0, 4, 1, 1, 8, 2, 0, 4, 0, 8, 1, 5, 2, 4, 1, 4, 7, 7, 2, 6, 0, 8, 7, 2, 5, 2, 6, 8] 
2.6.5 :016 > array.my_select(&block)
 => [2, 4, 3, 7, 1, 3, 1, 0, 6, 0, 4, 4, 3, 0, 3, 8, 6, 7, 8, 0, 5, 3, 2, 3, 2, 6, 5, 6, 2, 7, 6, 4, 3, 1, 0, 4, 0, 6, 6, 3, 0, 2, 0, 6, 8, 2, 2, 8, 3, 6, 4, 2, 1, 1, 4, 4, 3, 2, 8, 1, 6, 5, 4, 0, 6, 7, 5, 4, 8, 2, 2, 8, 0, 4, 1, 1, 8, 2, 0, 4, 0, 8, 1, 5, 2, 4, 1, 4, 7, 7, 2, 6, 0, 8, 7, 2, 5, 2, 6, 8] 
2.6.5 :017 > array.select(&block)
 => [2, 3, 1, 3, 1, 0, 0, 3, 0, 3, 0, 3, 2, 3, 2, 2, 3, 1, 0, 0, 3, 0, 2, 0, 2, 2, 3, 2, 1, 1, 3, 2, 1, 0, 2, 2, 0, 1, 1, 2, 0, 0, 1, 2, 1, 2, 0, 2, 2] 
2.6.5 :018 > 
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :020 > range = Range.new(5, 50)
2.6.5 :021 > range.select(&block)
 => [] 
2.6.5 :022 > range.my_select(&block)
 => 5..50 
2.6.5 :023 > 
```

File: lib/main.rb
Comment (maelfosso) [inline lib/main.rb]:
```rb
2.6.5 :025 > true_block = proc { |num| num <= 9 }
2.6.5 :026 > array
 => [2, 4, 3, 7, 1, 3, 1, 0, 6, 0, 4, 4, 3, 0, 3, 8, 6, 7, 8, 0, 5, 3, 2, 3, 2, 6, 5, 6, 2, 7, 6, 4, 3, 1, 0, 4, 0, 6, 6, 3, 0, 2, 0, 6, 8, 2, 2, 8, 3, 6, 4, 2, 1, 1, 4, 4, 3, 2, 8, 1, 6, 5, 4, 0, 6, 7, 5, 4, 8, 2, 2, 8, 0, 4, 1, 1, 8, 2, 0, 4, 0, 8, 1, 5, 2, 4, 1, 4, 7, 7, 2, 6, 0, 8, 7, 2, 5, 2, 6, 8] 
2.6.5 :027 > array.all?(&true_block)
 => true 
2.6.5 :028 > array.my_all?(&true_block)
 => false 
2.6.5 :029 > 
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :031 > false_block = proc { |num| num > 9 }
2.6.5 :032 > array
 => [2, 4, 3, 7, 1, 3, 1, 0, 6, 0, 4, 4, 3, 0, 3, 8, 6, 7, 8, 0, 5, 3, 2, 3, 2, 6, 5, 6, 2, 7, 6, 4, 3, 1, 0, 4, 0, 6, 6, 3, 0, 2, 0, 6, 8, 2, 2, 8, 3, 6, 4, 2, 1, 1, 4, 4, 3, 2, 8, 1, 6, 5, 4, 0, 6, 7, 5, 4, 8, 2, 2, 8, 0, 4, 1, 1, 8, 2, 0, 4, 0, 8, 1, 5, 2, 4, 1, 4, 7, 7, 2, 6, 0, 8, 7, 2, 5, 2, 6, 8] 
2.6.5 :033 > array.all?(&false_block)
 => false 
2.6.5 :034 > array.my_all?(&false_block)
 => true 
2.6.5 :035 > 
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :037 > false_array = [1, false, 'hi', []]
2.6.5 :038 > false_array.my_all?
 => true 
2.6.5 :039 > false_array.all?
 => false 
2.6.5 :040 >
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :042 > array
 => [2, 4, 3, 7, 1, 3, 1, 0, 6, 0, 4, 4, 3, 0, 3, 8, 6, 7, 8, 0, 5, 3, 2, 3, 2, 6, 5, 6, 2, 7, 6, 4, 3, 1, 0, 4, 0, 6, 6, 3, 0, 2, 0, 6, 8, 2, 2, 8, 3, 6, 4, 2, 1, 1, 4, 4, 3, 2, 8, 1, 6, 5, 4, 0, 6, 7, 5, 4, 8, 2, 2, 8, 0, 4, 1, 1, 8, 2, 0, 4, 0, 8, 1, 5, 2, 4, 1, 4, 7, 7, 2, 6, 0, 8, 7, 2, 5, 2, 6, 8] 
2.6.5 :043 > array.my_all?(Integer)
 => false 
2.6.5 :044 > array.all?(Integer)
 => true 
2.6.5 :045 > 
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :047 > numbers = [1, 2i, 3.14]
2.6.5 :048 > numbers.all?(Numeric)
 => true 
2.6.5 :049 > numbers.my_all?(Numeric)
 => false 
2.6.5 :050 >
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :052 > words = %w[dog door rod blade]
2.6.5 :053 > words.all?(/d/)
 => true 
2.6.5 :054 > words.my_all?(/d/)
 => false 
2.6.5 :055 >
```

Comment (maelfosso) [reply lib/main.rb]:
```rb
2.6.5 :062 > arr = []
2.6.5 :063 > 5.times { arr << 3 }
 => 5 
2.6.5 :064 > arr.my_all?(3)
 => false 
2.6.5 :065 > arr.all?(3)
 => true 
2.6.5 :066 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for making the suggested changes
However, there are still some bugs to fix before moving forwards
Kindly read the comments left below. 

Once fixing the bugs, submit a new code review request
Thanks

Mael FOSSO

File: README.md
Comment (kitpao) [inline README.md]: - [x] **OPTIONAL** You may want to fix the typo "enumearble"

File: lib/main.rb
Comment (kitpao) [inline lib/main.rb]: - [x] Redundant self is an easy linter to fix, you just have to remove your `self`, I'll guide you in the following comments

File: lib/main.rb
Comment (kitpao) [inline lib/main.rb]: - [x] Here, you can put self into a variable and then use it normally

    ```ruby
      object = self
      items = to_a if object.class == Hash || Range
    ```

File: lib/main.rb
Comment (kitpao) [inline lib/main.rb:19]: - Here `self` is okay since there is no other method with it ✔️

File: lib/main.rb
Comment (kitpao) [inline lib/main.rb]: - [x] Please delete this linter's enable and disable for Style/RedundantSelf, if you have any doubt, you can contact me or a full time TSE [here](https://microverse.zendesk.com/hc/en-us/articles/360026625973-How-can-I-get-help-from-TSEs-)

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @mke2111 

Good job trying to fix the failing tests indicated by previous TSE 👍 However we still have some issues, please go below to understand them better 😄 

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

### Methods:

#### #my_select

- [x] Notice that your select method should return the filtered array
```ruby
p [1, 2, 3].select{ |x| x >= 2 }              # original returns [2,3]
p [1, 2, 3].my_select{ |x| x >= 2 }           #yours returns the initial array [1,2,3]
```
**Hint:** Pay attention to what you are returning in your method

#### #my_all?
You have a bug in this line that is causing almost all the failing tests:
```ruby
my_each { |item| return false if yield(item) }
# Are you sure you want to return something when the yield is true?
# when yield is true, you want to keep on going and keep searching for a condition that makes you want to stop and return something
```
  ##### Sample failing tests because of this issue:
  - [x] `[1, 2].my_all? { |x| x > 0}` should be true because the condition is true for all the methodsnot all of the elements are true. Your method returns false
  
 - [x] `[1, false].my_all?` should be false because not all of the elements are true. Your method returns true
  ##### After you fix the previous issue, you will have 1 failing test: 
  - [x] `p [true, true, true].my_all?` should be true

📝 Please, do not forget the inline comments below 🔽 

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

#### Hello @mke2111, Great job! 👏  

You've done an excellent job so far 👌. Let's just fix #my_all? method and you will be good to go.

- [x] Your `#my_all?` method should return true when none of the collection members are false or nil' e.g `[1, true, 'hi', []].my_all?` should return `true`

Make this one last change and you will be on your way to the next project.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: lib/main.rb
Comment (maelfosso) [inline lib/main.rb]:
```rb
2.6.5 :003 > false_array = [1, false, 'hi', []]
2.6.5 :004 > false_array.all?
 => false 
2.6.5 :005 > false_array.my_all?
 => true 
2.6.5 :006 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for making the previous changes.
However, there is a bug to fix.
Read the comment left below. 

Once fixing the bug submit a new code review request.

Mael FOSSO

File: lib/main.rb
Comment (peterrobert) [inline lib/main.rb]: when no block or argument is given it should return true when none of the collection members are false or nil. Kindly check below.
### YOURS
`true_array = [1, true, 'hi', []]
puts true_array.my_all?
1
true
hi
`
### ORIGINAL
`true_array = [1, true, 'hi', []]
puts true_array.all?
true
`

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :stop_sign: :recycle: 
### HELLO :wave: @mke2111 VERY NICE JOB. YOU HAVE TRIED TO MAKE THE CHANGES AS REQUESTED BY THE PREVIOUS TSE. :+1: THERE IS STILL SOME CHANGES TO BE DONE KINDLY WORK ON THE FOLLOWING :arrow_down: 

**IF YOU HAVE ANY QUESTION KINDLY REACH ME ON SLACK @PETERROBERT**

### HAPPY CODING :computer:

**Review by peterrobert — APPROVED** [review]
### APPROVED :+1: :100: :1st_place_medal: 
### VERY NICE WORK :100: GREAT PROGRESS :+1: ON TO THE NEXT PROJECT YOUR WORK IS APPROVED

### HAPPY CODING :computer:

**Pull request merged**
